qualtran/bloqs/mod_arithmetic/mod_multiplication.py

```diff
@@ -266,3 +273,395 @@ def _modmul_symb() -> CModMulK:
+
+        if isinstance(self.window_size, int) and isinstance(self.bitsize, int):
+            assert self.window_size <= self.bitsize
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        num_windows = (
+            self.bitsize + self.window_size - 1
+        ) // self.window_size  # = ceil(self.bitsize/self.window_size)
+        return Signature(
+            [
+                Register('x', QMontgomeryUInt(self.bitsize)),
+                Register('y', QMontgomeryUInt(self.bitsize)),
+                Register('target', QMontgomeryUInt(self.bitsize)),
+                Register('qrom_indices', QMontgomeryUInt(num_windows * self.window_size)),
+                Register('reduced', QBit()),
+            ]
+        )
+
+    @cached_property
+    def _lookup_qrom(self) -> QROM:
+        if is_symbolic(self.window_size) or is_symbolic(self.mod):
+            return QROM(
+                [Shaped((2**self.window_size,))],
+                selection_bitsizes=(self.window_size,),
+                target_bitsizes=(self.window_size + self.bitsize,),
+            )
+        inv_mod = pow(self.mod, 2 ** (self.window_size - 1) - 1, 2**self.window_size)
+        N = 2**self.window_size
+        data = (-np.arange(N) * inv_mod) % N
+        data *= self.mod
+        return QROM(
+            [data],
+            selection_bitsizes=(self.window_size,),
+            target_bitsizes=(self.window_size + self.bitsize,),
+        )
+
+    def _apply_window(
+        self, bb, x_arr, y: Soquet, target_arr, qrom_index: Soquet, qrom_target: Soquet
+    ):
+        if is_symbolic(self.window_size):
+            raise DecomposeNotImplementedError(f'symbolic decomposition not supported for {self}')
+        for i in range(self.window_size):
+            z = bb.join(target_arr[-self.bitsize - 1 - i : len(target_arr) - i])
+            x_arr[i], y, z = bb.add(
+                CAdd(QMontgomeryUInt(self.bitsize), QMontgomeryUInt(self.bitsize + 1)),
+                ctrl=x_arr[i],
+                a=y,
+                b=z,
+            )
+            z_arr = bb.split(z)
+            target_arr[-self.bitsize - 1 - i : len(target_arr) - i] = z_arr
+
+        m = bb.join(target_arr[-self.window_size :], QMontgomeryUInt(self.window_size))
+        m, qrom_index = bb.add(Xor(QMontgomeryUInt(self.window_size)), x=m, y=qrom_index)
+        target_arr[-self.window_size :] = bb.split(m)
+
+        qrom = self._lookup_qrom
+        qrom_index, qrom_target = bb.add(qrom, selection=qrom_index, target0_=qrom_target)
+        z = bb.join(target_arr)
+        qrom_target, z = bb.add(
+            Add(QMontgomeryUInt(self.bitsize + self.window_size)), a=qrom_target, b=z
+        )
+        qrom_index, qrom_target = bb.add(qrom.adjoint(), selection=qrom_index, target0_=qrom_target)
+        target_arr = bb.split(z)
+        target_arr = np.roll(target_arr, self.window_size)
+
+        return x_arr, y, target_arr, qrom_index, qrom_target
```

Comment:
code smell: usually when you feel the need to factor out a function in python-land it means you should probably factor out a bloq in quantum land